### PR TITLE
Avm now handles "pure" balance transfers

### PR DESCRIPTION
**Pull request by aionick**
## Description

- Formerly, transactions that simply transferred balance to an address that is not a contract were executed by the Fvm. Now they are executed by the Avm.
- This now mean that a CALL transaction is avm-bound if: the call is to an avm contract _or to a non-contract address_.
- Some minor refactorings to the `BulkExecutor`; namely, removing the unnecessary `isTransactionForFvm()` method which must simply return `!isTransactionForAvm()`.
- There is a discrepancy between vm's handling the pay miner fees & refund sender. The Avm does this whereas the Fvm does not (will open an issue for this soon). This explains the other logic change in the `BulkExecutor`, which only does these actions if the transaction is not avm-bound.

**The actual changes to logic are in commit b177d64 (the other commit, b61bc8c contains tests).**

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [ ] New feature.
- [x] Enhancement.
- [ ] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

- A number of consensus tests were added before making the change, in the first commit here, to ensure consensus was not broken. The test harness also contains some tests on balance transfers (soon to be pushed)

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [x] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [x] Any dependent changes have been made.

### CODE REVIEW

File: modVM/src/org/aion/vm/BulkExecutor.java [hunk 1]
```diff
@@ -124,7 +123,7 @@ public BulkExecutor(
      * @param logger The logger.
      * @param work The post-execution work to apply after each transaction is run.
      */
-    public BulkExecutor (
+    public BulkExecutor(
```
Comment (AlexandraRoatis) [inline modVM/src/org/aion/vm/BulkExecutor.java]: formatting

File: modVM/src/org/aion/vm/BulkExecutor.java [hunk 2]
```diff
@@ -343,25 +345,25 @@ private void updateRepository(
...
             // Refund energy if transaction was successfully or reverted.
-            if (result.getResultCode().isSuccess() || result.getResultCode().isRevert()) {
-                if (!transactionIsForAionVirtualMachine(tx)) {
+            if (!transactionIsForAionVirtualMachine(tx)) {
+                if (result.getResultCode().isSuccess() || result.getResultCode().isRevert()) {
                     track.addBalance(tx.getSenderAddress(), summary.getRefund());
                 }
-            }
 
-            tx.setNrgConsume(computeEnergyUsed(tx.getEnergyLimit(), result));
```
Comment (AlexandraRoatis) [inline modVM/src/org/aion/vm/BulkExecutor.java]: where is this done now?

Comment (aionick) [reply modVM/src/org/aion/vm/BulkExecutor.java]: It just moved up a few lines, before the if statement, since it is not conditional

**Comment by aionick** [comment]
I won't have time to do the assertThat change, I'll try to keep it in mind in the future

File: modVM/src/org/aion/vm/BulkExecutor.java [hunk 3]
```diff
@@ -157,45 +156,47 @@ public BulkExecutor (
...
+                            new KernelInterfaceForFastVM(
                                     this.repositoryChild.startTracking(),
                                     this.allowNonceIncrement,
                                     this.isLocalCall);
                     virtualMachineForNextBatch =
-                            VirtualMachineProvider.getVirtualMachineInstance(VM.AVM, vmKernel);
+                            VirtualMachineProvider.getVirtualMachineInstance(VM.FVM, vmKernel);
                     nextBatchToExecute =
-                            fetchNextBatchOfTransactionsForAionVirtualMachine(currentIndex);
+                            fetchNextBatchOfTransactionsForFastVirtualMachine(currentIndex);
```
Comment (AionJayT) [inline modVM/src/org/aion/vm/BulkExecutor.java]: reason for reversing avm fvm order in the if-else condition

Comment (aionick) [reply modVM/src/org/aion/vm/BulkExecutor.java]: Just seems more natural to do `if (condition)` instead of `if (!condition)`

**Comment by AionJayT** [comment]
please fix the conflict. Thanks

File: modVM/src/org/aion/vm/BulkExecutor.java (hunk 1 above)
Comment (aionick) [reply modVM/src/org/aion/vm/BulkExecutor.java]: I don't think that's a formatting mistake. That's the result of applying the formatter, lol.

Comment (AlexandraRoatis) [reply modVM/src/org/aion/vm/BulkExecutor.java]: ups, sorry about that!

**Pull request merged**